src/metadata_directory/main.mo

```diff
@@ -14,83 +15,70 @@ actor class () = self {
     description : Text;
     build_instructions : Text;
     latest_update_timestamp : Nat64;
+    created_timestamp : Nat64;
   };
 
-  /// Stores all the metadata.
-  ///
-  /// 1st dimension - metadata directories divided by the principal.
-  ///
-  /// 2nd dimension - wasm module metadata records for a principal.
-  stable var storage : Vec.Vector<Vec.Vector<WasmMetadata>> = Vec.new();
-
-  /// Map principal to index of the directory managed by the principal.
-  let storage_map = RBTree.RBTree<Principal, Nat>(Principal.compare);
+  public type PrincipalRecord = {
+    var wasm_metadata_storage : OrderedMap.Map<Blob, WasmMetadata>;
+  };
 
-  /// Map (principal, module_hash) to index of wasm module metadata record.
-  let wasm_modules_map = RBTree.RBTree<(Principal, Blob), Nat>(
-    func(
-      a : (Principal, Blob),
-      b : (Principal, Blob),
-    ) {
-      switch (Principal.compare(a.0, b.0)) {
-        case (#equal) Blob.compare(a.1, b.1);
-        case (value) value;
-      };
-    }
-  );
+  let principalMap = OrderedMap.Make<Principal>(Principal.compare);
+  let blobMap = OrderedMap.Make<Blob>(Blob.compare);
 
-  stable var stable_storage_map = storage_map.share();
-  stable var stable_wasm_modules_map = wasm_modules_map.share();
+  stable var storage : OrderedMap.Map<Principal, PrincipalRecord> = principalMap.empty<PrincipalRecord>();
 
   //
   // API for Metadata directory management
   //
 
   public shared query ({ caller }) func get_wasm_metadata() : async [WasmMetadata] {
-    let ?index = storage_map.get(caller) else return [];
-    Vec.get(storage, index) |> Vec.toArray(_);
+    let ?pr = principalMap.get(storage, caller) else return [];
+    pr.wasm_metadata_storage
+    |> blobMap.vals(_)
+    |> Iter.toArray(_);
   };
 
   public shared ({ caller }) func add_wasm_metadata(module_hash : Blob, description : ?Text, build_instructions : ?Text) : async () {
-    let wasm_modules = switch (storage_map.get(caller)) {
-      case (?index) Vec.get(storage, index);
+    let pr = switch (principalMap.get(storage, caller)) {
```

Comment:
In this function `add_wasm_metadata` (not in `update_wasm_metadata`) I think the arguments should be 
```
add_wasm_metadata(module_hash : Blob, description : Text, build_instructions : Text)
```
This means no option before Text. If the user wants he can supply `""` (empty string). It is just as easy as supplying `null`. And is more explicit.